src/bijectors/coupling_layer.jl

```diff
@@ -0,0 +1,244 @@
+parts of the vector you want to apply the bijector to.
+"""
+function PartitionMask(n::Int, indices)
+    indices_2 = [i for i in 1:n if i ∉ indices]
+
+    # sparse arrays <3
+    A_1 = spzeros(n, length(indices));
+    A_2 = spzeros(n, length(indices_2));
+
+    # Like doing:
+    #    A[1, 1] = 1.0
+    #    A[3, 2] = 1.0
+    for (i, idx) in enumerate(indices)
+        A_1[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = 1.0
+    end
+
+    return PartitionMask(A_1, A_2, spzeros(n, 0))
+end
+PartitionMask(x::AbstractVector, indices) = PartitionMask(length(x), indices)
+
+"""
+    combine(m::PartitionMask, x_1, x_2, x_3)
+
+Combines `x_1`, `x_2`, and `x_3` into a single vector.
+"""
+@inline combine(m::PartitionMask, x_1, x_2, x_3) = m.A_1 * x_1 .+ m.A_2 * x_2 .+ m.A_3 * x_3
+
+"""
+    partition(m::PartitionMask, x)
+
+Partitions `x` into 3 disjoint subvectors.
+"""
+@inline partition(m::PartitionMask, x) = (transpose(m.A_1) * x, transpose(m.A_2) * x, transpose(m.A_3) * x)
+
+
+# CouplingLayer
+
+"""
+    CouplingLayer{B, M, F}(mask::M, θ::F)
+
+Implements a coupling-layer as defined in [1].
+
+# Examples
+```julia-repl
+julia> m = PartitionMask(3, [1], [2]) # <= going to use x[2] to parameterize transform of x[1]
+PartitionMask{SparseArrays.SparseMatrixCSC{Float64,Int64}}(
+  [1, 1]  =  1.0, 
+  [2, 1]  =  1.0, 
+  [3, 1]  =  1.0)
+
+julia> cl = CouplingLayer(Shift, m, identity) # <= will do `y[1:1] = x[1:1] + x[2:2]`;
+
+julia> x = [1., 2., 3.];
+
+julia> cl(x)
+3-element Array{Float64,1}:
+ 3.0
+ 2.0
+ 3.0
+```
+
+# References
+[1] Kobyzev, I., Prince, S., & Brubaker, M. A., Normalizing flows: introduction and ideas, CoRR, (),  (2019). 
+"""
+struct CouplingLayer{B, M, F} <: Bijector{1} where {B, M <: PartitionMask, F}
+    mask::M
+    θ::F
+end
+
+CouplingLayer(B, mask::M, θ::F) where {M, F} = CouplingLayer{B, M, F}(mask, θ)
+function CouplingLayer(B, θ, n::Int)
+    idx = Int(floor(n / 2))
+    return CouplingLayer(B, PartitionMask(n, 1:idx), θ)
+end
+
+function CouplingLayer(cl::CouplingLayer{B}, mask::PartitionMask) where {B}
+    return CouplingLayer(B, mask, cl.θ)
+end
+
+"Returns the constructor of the coupling law."
+coupling(cl::CouplingLayer{B}) where {B} = B
+
+"Returns the coupling law constructed from `x`."
+function couple(cl::CouplingLayer{B}, x::AbstractVector) where {B}
+    # partition vector using `cl.mask::PartitionMask`
+    x_1, x_2, x_3 = partition(cl.mask, x)
+
+    # construct bijector `B` using θ(x₂)
+    b = B(cl.θ(x_2))
+
+    return b
+end
+
+function (cl::CouplingLayer{B})(x::AbstractVector) where {B}
+    # partition vector using `cl.mask::PartitionMask`
+    x_1, x_2, x_3 = partition(cl.mask, x)
+
+    # construct bijector `B` using θ(x₂)
+    b = B(cl.θ(x_2))
+
+    # recombine the vector again using the `PartitionMask`
+    return combine(cl.mask, b(x_1), x_2, x_3)
+end
+function (cl::CouplingLayer{B})(x::AbstractMatrix) where {B}
```

Comment:
Same question for logabsdetjac and those for inverse

Reply:
It isn't absolutely necessary, but otherwise we need `cl.θ` to be able to handle both `AbstractVector` and `AbstractMatrix` i.e. more work for the user.

Worth mentioning that this was written _before_ we introduced dimensionality, so these might not be necessary anymore!:)